### PR TITLE
feat: add storage provider field to support s3 and local storage

### PR DESCRIPTION
**Pull request by zyy17**

### CODE REVIEW

File: cmd/initializer/main.go [hunk 1]
```diff
@@ -43,6 +54,16 @@ func main() {
 	pflag.StringVar(&opts.datanodeServiceName, "datanode-service-name", "", "the name of datanode service")
 	pflag.StringVar(&opts.namespace, "namespace", "", "the cluster namespace")
 
+	// Handle storage options.
+	pflag.StringVar(&opts.storageType, "storage-type", "", "the storage type")
+	pflag.StringVar(&opts.localStorageDir, "local-storage-dir", "", "the directory of local storage")
+	pflag.StringVar(&opts.bucket, "bucket", "", "the bucket of s3 storage")
+	pflag.StringVar(&opts.prefix, "prefix", "", "the prefix of s3 storage")
+	pflag.StringVar(&opts.accessKeyID, "access-key-id", "", "the access key id of s3 storage")
+	pflag.StringVar(&opts.secretAccessKey, "secret-access-key", "", "the access key id of s3 storage")
```
Comment (daviderli614) [inline cmd/initializer/main.go:63]: Same explain accessKeyID and secretAccessKey

File: config/samples/s3-storage/s3-credentials.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: Secret
+type: Opaque
+metadata:
+  name: s3-credentials
+  namespace: default
+data:
+  access-key-id: "your-access-key-id"
+  secret-access-key: "your-secret-access-key"
```
Comment (daviderli614) [inline config/samples/s3-storage/s3-credentials.yaml:9]: need have licenses in file ?

File: controllers/greptimedbcluster/deployers/datanode.go [hunk 3]
```diff
@@ -436,11 +452,36 @@ func (d *DatanodeDeployer) generatePodTemplateSpec(cluster *v1alpha1.GreptimeDBC
...
+		if cluster.Spec.StorageProvider.Local != nil {
+			storageArgs := []string{
+				"--storage-type", "Local",
+				"--local-storage-dir", cluster.Spec.StorageProvider.Local.Directory,
+			}
+			initializer.Args = append(initializer.Args, storageArgs...)
+		}
 
-	return podTemplateSpec
+		if cluster.Spec.StorageProvider.S3 != nil {
```
Comment (daviderli614) [inline controllers/greptimedbcluster/deployers/datanode.go:466]: if localstorage != nil && s3storage != nil, what to do.

File: controllers/greptimedbcluster/greptimedbcluster_controller.go [hunk 4]
```diff
@@ -276,6 +276,24 @@ func (r *Reconciler) validate(ctx context.Context, cluster *v1alpha1.GreptimeDBC
...
+	if cluster.Spec.StorageProvider != nil {
+		if cluster.Spec.StorageProvider.S3 != nil {
+			storageCredentialsSecret := &corev1.Secret{}
+			err := r.Get(ctx, types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.StorageProvider.S3.SecretName}, storageCredentialsSecret)
+			if err != nil {
+				return fmt.Errorf("get storage credentials secret '%s' failed, error: '%v'", cluster.Spec.StorageProvider.S3.SecretName, err)
+			}
+
+			if _, ok := storageCredentialsSecret.Data[deployers.AccessKeyIDSecretKey]; !ok {
+				return fmt.Errorf("credentials secret '%s' does not contain key '%s'", cluster.Spec.StorageProvider.S3.SecretName, deployers.AccessKeyIDSecretKey)
```
Comment (daviderli614) [inline controllers/greptimedbcluster/greptimedbcluster_controller.go:288]: SecretAccessKeySecretKey, AccessKeyIDSecretKey have same error log

File: controllers/greptimedbcluster/deployers/datanode.go (hunk 3 above)
Comment (zyy17) [reply controllers/greptimedbcluster/deployers/datanode.go:466]: @daviderli614 I already add the validation in `validate()`, if the case happen, controller will reject the creation.

**Review by daviderli614 — APPROVED** [review]
LGTM

**Pull request merged**